### DISCUSSION TITLE
How are TLS implemented for Wasmer

### DISCUSSION THREAD

**Discussion by yizhuoliang** [wasmerio/wasmer#4814] [General]
Traditionally, TLS are managed by libpthread in the user space, but it seems like Wasmer provisions TLS external to the WASM module? Then does these TLS variables reside inside the instance's Linear Memory or in somewhere external?

Also, what is the TLS model that's been used? `local-dynamic` or `global-dynamic`?

**Comment by xdoardo** [comment]
We used to have TLS managed by runtime with associated syscalls, but a while ago we moved to use `libpthread`, so TLS is now managed by that library. If you want to know more, check out the code in the [wasix-libc repo](https://github.com/wasix-org/wasix-libc)!

Regarding the particular model used, I don't quite remember, but it should be clear from the repo above. If you have any specific need, feel free to ask!
